server/game/gameSystems/DiscardFromDeckSystem.ts

```diff
@@ -0,0 +1,89 @@
+import { AbilityContext } from '../core/ability/AbilityContext';
+import { Card } from '../core/card/Card';
+import { EventName, GameStateChangeRequired } from '../core/Constants';
+import { IPlayerTargetSystemProperties, PlayerTargetSystem } from '../core/gameSystem/PlayerTargetSystem';
+import Player from '../core/Player';
+import { DiscardSpecificCardSystem } from './DiscardSpecificCardSystem';
+import * as Contract from '../core/utils/Contract';
+
+export interface IDiscardFromDeckProperties extends IPlayerTargetSystemProperties {
+    amount?: number;
+}
+
+export class DiscardFromDeckSystem<TContext extends AbilityContext = AbilityContext> extends PlayerTargetSystem<TContext, IDiscardFromDeckProperties> {
+    public override readonly name = 'discardFromDeck';
+    public override readonly eventName = EventName.OnDiscardFromDeck;
+
+    protected override defaultProperties: IDiscardFromDeckProperties = {
+        amount: 1
+    };
+
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    public override eventHandler(_event): void { }
+
+    // public eventHandler(event): void {
+    //     for (let i = 0; i < event.amount; i++) {
+    //         const topCard = event.player.getTopCardOfDeck();
+    //         event.player.moveCard(topCard, Location.Discard);
+    //     }
+    // }
+
+    public override getEffectMessage(context: TContext): [string, any[]] {
+        const properties = this.generatePropertiesFromContext(context);
+        return ['discard ' + (properties.amount + (properties.amount > 1 ? ' cards' : ' card') + ' from deck'), []];
+    }
+
+    public override canAffect(player: Player, context: TContext, additionalProperties = {}, mustChangeGameState = GameStateChangeRequired.None): boolean {
+        const properties = this.generatePropertiesFromContext(context, additionalProperties);
+
+        const availableDeck = player.drawDeck || []; // Without this | [], sometimes this is undefined, causing issues below
+
+        if (mustChangeGameState !== GameStateChangeRequired.None && (availableDeck.length === 0 || properties.amount === 0)) {
+            return false;
+        }
+
+        if ((properties.isCost || GameStateChangeRequired.MustFullyResolve) && availableDeck.length < properties.amount) {
+            return false;
+        }
+
+        if (!super.canAffect(player, context, additionalProperties)) {
+            return false;
+        }
+
+        return properties.amount !== 0 && super.canAffect(player, context);
```

Comment:
Looks we can shortcut pretty early if amount === 0, maybe let's put that as its own if check pretty high in the function? That way we could remove the check from e.g. line 41 so we're not checking it twice